src/odoc/bin/main.ml

```diff
@@ -476,28 +476,44 @@ module Compile_impl = struct
     in
     Term.info "compile-impl" ~docs ~doc
 end
+let ( >>= ) = Result.bind
 
 module Indexing = struct
-  let output_file ~dst =
-    match dst with
-    | Some file -> Fs.File.of_string file
-    | None -> Fs.File.of_string "index.json"
-
-  let index dst warnings_options inputs_in_file inputs =
-    let output = output_file ~dst in
-    match (inputs_in_file, inputs) with
-    | [], [] ->
+  let output_file ~dst marshall =
+    match (dst, marshall) with
+    | Some file, `JSON when not (Fpath.has_ext "json" (Fpath.v file)) ->
+        Error
+          (`Msg
+            "When generating a json index, the output must have a .json file \
+             extension")
+    | Some file, `Marshall when not (Fpath.has_ext "odoc-index" (Fpath.v file))
+      ->
+        Error
+          (`Msg
+            "When generating a binary index, the output must have a \
+             .odoc-index file extension")
+    | Some file, _ -> Ok (Fs.File.of_string file)
+    | None, `JSON -> Ok (Fs.File.of_string "index.json")
+    | None, `Marshall -> Ok (Fs.File.of_string "index.odoc-index")
+
+  let index dst json warnings_options includes_rec inputs_in_file inputs =
+    let marshall = if json then `JSON else `Marshall in
+    output_file ~dst marshall >>= fun output ->
+    match (includes_rec, inputs_in_file, inputs) with
+    | [], [], [] ->
         Result.Error
           (`Msg
-            "At least one of --file-list or an .odocl file must be passed to \
-             odoc compile-index")
-    | _ -> Indexing.compile ~output ~warnings_options inputs_in_file inputs
-
+            "At least one of --file-list or --include-rec or an .odocl file \
+             must be passed to odoc compile-index")
```

Comment:
This is already checked in `Indexing.compile` but in a slightly different way: If `--file-list` points to an empty file, this passes but `Indexing.compile` has an other check that do not.

Why shouldn't making an empty index be allowed ? For example, drivers might want to make an index for every packages, even empty or bugous packages. This restriction increases complexity in drivers.

Reply:
Good point.